File: src/main/java/no/fintlabs/mapping/FormattingUtilsService.java
package no.fintlabs.mapping;

import org.apache.commons.lang3.StringUtils;
import org.springframework.stereotype.Service;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

@Service
public class FormattingUtilsService {

    public String formatEmail(String email) {
        return email.trim().toLowerCase();
    }

    public String extractEmailDomain(String email) {
        if (email == null || !email.contains("@")) {
            return "Invalid email";
        }
        return email.substring(email.indexOf("@") + 1).toLowerCase();
    }


    public String formatKommunenavn(String kommunenavn) {
        if (kommunenavn == null || kommunenavn.isEmpty()) {
            return kommunenavn;
        }

        kommunenavn = kommunenavn.replaceAll("\\s+", " ").toLowerCase();

        Matcher matcher = Pattern.compile("[^- ]+|[- ]").matcher(kommunenavn);
        StringBuilder result = new StringBuilder();

        while (matcher.find()) {
            String token = matcher.group();
            if (token.equals(" ") || token.equals("-")) {
                result.append(token);
            } else {
                result.append(StringUtils.capitalize(token));
            }
        }

        return result.toString();
    }
}

File: src/main/kotlin/no/novari/flyt/egrunnerverv/gateway/instance/mapping/FormattingUtilsService.kt
package no.novari.flyt.egrunnerverv.gateway.instance.mapping

import org.springframework.stereotype.Service
import java.util.Locale

@Service
class FormattingUtilsService {
    private val locale: Locale = Locale.forLanguageTag("nb-NO")

    fun formatEmail(email: String): String = email.trim().lowercase()

    fun extractEmailDomain(email: String): String? =
        email
            .substringAfter('@', "")
            .trim()
            .lowercase()
            .takeIf { it.isNotBlank() }

    fun formatKommunenavn(kommunenavn: String?): String? {
        if (kommunenavn.isNullOrBlank()) {
            return kommunenavn
        }

        val normalized = kommunenavn.replace("\\s+".toRegex(), " ").trim()
        return capitalizeFully(normalized)
    }

    fun capitalizeFully(input: String): String {
        val delimiters = setOf(' ', '-')

        val result = StringBuilder(input.length)
        var capitalizeNext = true

        for (char in input.lowercase(locale)) {
            if (capitalizeNext && char.isLetter()) {
                result.append(char.titlecase(locale))
                capitalizeNext = false
            } else {
                result.append(char)
                capitalizeNext = char in delimiters
            }
        }

        return result.toString()
    }
}
